[WIP] Workaround to overflow issue in power consumption

paulolieuthier opened the pull request:
This is a work in progress.

The idea is to use a different thread to collect the power consumption information regularly, as we know the overflow might happen after 60 seconds. As we discussed, we know the overflow can also happen before 60 seconds, so this is still an issue.

Issues to solve:
- [x] overflow
- [x] there is a data race in the `energy_sum` variable
- [ ] `criterion_getenergypacket` might be called during a sleep period of the second thread

@luisgabriel  do you think there is a better way? What about the isses pointed out?

paulolieuthier committed:
Workaround to overflow issue in power consumption

luisgabriel commented on cbits/energy-rapl.c:
```diff
@@ -1,28 +1,40 @@
+    {
+        sleep(15);
+        struct rapl_raw_power_counters rpc;
+        rapl_get_raw_power_counters(fd_msr, &r_units, &rpc);
```
Consider using [rapl_get_power_diff()](https://github.com/anyc/librapl/blob/5d823570786595b3d527ed97a8e7a57e698adceb/rapl.h#L112) here. You can accumulate just the diff and if an overflow happens (diff < 0) you can discard the current one.

paulolieuthier replied:
Awesome. I was reading [perf's source](https://github.com/torvalds/linux/blob/master/arch/x86/kernel/cpu/perf_event_intel_rapl.c#L170 to achieve that. Just leaving here for future reference.

paulolieuthier replied:
Done.

luisgabriel commented:
Initially I was thinking in start the thread only when we're actually collecting energy data (where we are calling the first `getTime` in the current codebase) but I'm not sure if this is the best solution since there is an overhead to create/destroy threads.

paulolieuthier committed:
Update our rapl code from upstream

That was needed for `rapl_get_power_diff`. The previous code had a lot
of commented lines, for some reason. I'm not sure if it's a problem.

paulolieuthier committed:
Use `rapl_get_power_diff` to workaround overflow

paulolieuthier commented:
> Initially I was thinking in start the thread only when we're actually collecting energy data (where we are calling the first getTime in the current codebase) but I'm not sure if this is the best solution since there is an overhead to create/destroy threads.

I've been thinking, that might be a better ideia indeed. I don't think the overhead of creating the new thread will interfere in the real measurement, as the benchmark is run after `getTime` returns (the thread is already created).

paulolieuthier committed:
Use pthread's mutex to eliminate data race

paulolieuthier commented:
Closing this because the sigalarm method will be used.
